src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -1994,23 +1994,57 @@ impl Coordinator {
 
     /// Checks to see if the session needs a real time recency timestamp and if so returns
     /// a future that will return the timestamp.
-    pub(super) fn recent_timestamp(
+    pub(super) async fn recent_timestamp(
         &self,
         session: &Session,
         source_ids: impl Iterator<Item = GlobalId>,
-    ) -> Option<BoxFuture<'static, Timestamp>> {
+    ) -> Result<Option<BoxFuture<'static, Result<Timestamp, StorageError<Timestamp>>>>, AdapterError>
+    {
+        let vars = session.vars();
+
         // Ideally this logic belongs inside of
         // `mz-adapter::coord::timestamp_selection::determine_timestamp`. However, including the
         // logic in there would make it extremely difficult and inconvenient to pull the waiting off
         // of the main coord thread.
-        if session.vars().real_time_recency()
-            && session.vars().transaction_isolation() == &IsolationLevel::StrictSerializable
+        let r = if vars.real_time_recency()
+            && vars.transaction_isolation() == &IsolationLevel::StrictSerializable
             && !session.contains_read_timestamp()
         {
-            Some(self.controller.recent_timestamp(source_ids))
+            // Find all dependencies transitively because we need to ensure that
+            // RTR queries determine the timestamp from the sources' (i.e.
+            // storage objects that ingest data from external systems) remap
+            // data. We "cheat" a little bit and filter out any IDs that aren't
+            // user objects because we know they are not a RTR source.
+            let mut to_visit = VecDeque::from_iter(source_ids.filter(GlobalId::is_user));
+            // If none of the sources are user objects, we don't support RTR.
+            if to_visit.is_empty() {
+                return Ok(None);
+            }
+
+            let mut timestamp_objects = BTreeSet::new();
+
+            while let Some(id) = to_visit.pop_front() {
+                timestamp_objects.insert(id);
+                to_visit.extend(
+                    self.catalog()
+                        .get_entry(&id)
+                        .uses()
+                        .into_iter()
+                        .filter(|id| !timestamp_objects.contains(id) && id.is_user()),
+                );
+            }
+
+            let r = self
+                .controller
+                .recent_timestamp(timestamp_objects, *vars.statement_timeout())
```

Comment:
`statement_timeout` may be wrong here? Its docs say

> Sets the maximum allowed duration of INSERT...SELECT, UPDATE, and DELETE operations.

I'm not opposed to leaving it as is because we just need some timeout, and this isn't a bad one and there's not an obvious alternative.

Reply:
Yeah, I think you're right and this should just have its own timeout.